src/ray/raylet/worker_pool/cache_policy.h

```diff
@@ -0,0 +1,170 @@
+// Copyright 2017 The Ray Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <cstddef>
+#include <list>
+#include <memory>
+#include <unordered_set>
+
+#include "absl/container/flat_hash_set.h"
+#include "ray/common/id.h"
+#include "ray/raylet/worker.h"
+
+//#include "gtest/gtest.h"
+
+/*
+
+Goal:
+    * Implement the current policy behind new interface.
+        Prestarting is behind the flag -- so "desired" is zero? and prestart schedules
+creation
+
+Worker pool policy
```

Comment:
i'll just think out loud in the PR :)

### step 1
the **_policy_** should subscribe to all relevant **_events_** and makes decision of the **_target worker pool state_**. and the worker pool should consume the **_target worker pool state_** and adjust the current _**workers**_ to match the target.

let's start by assuming worker can be created with zero overhead, (this implied the decision the policy made can be immediately reflected by the worker pool), and worker creation never fails.
what are the potential events that the policy might want to know to the worker pool's target state?

- worker_pool_started (implicitly, cluster runtime env)
- job submitted (implicitly, job runtime env)
- job finished
- a cached worker being leased (to run a task/actor with a runtime env)
- a new worker being created to fulfill a pending request
- a worker being returned
- a worker normally exited
- a worker unexpected exited

basically, the policy can react to each of these event, look up the workers state in the form of (WORKER_ID -> (RuntimeEnv, Resources, Leased-client), and then create or delete cached workers, and wait for next event kicks in.

### step 2
now let's relax our assumption: worker creation could take a while, and it might fail. essentially we need track additional events, and store states accordingly:

events:
- worker creation succeeded
- worker creation failed

state:
the policy need also to track additional state for each worker with an **_unique_worker_id_**, and the worker's status (PENDING, SUCCESS_CREATED, FAILURE), so that when it receives worker creation success/failure events, it can update its internal state accordingly.

With this, whenever the policy receives events, it will look up the workerpool state (unique_worker_id -> (status, runtime_env, resources, leaser...), and make a decision to start, kill or cancel pending start workers.